processor/cumulativetodeltaprocessor/processor.go

```diff
@@ -53,64 +58,92 @@ func (ctdp *cumulativeToDeltaProcessor) Start(context.Context, component.Host) e
+		ilms.RemoveIf(func(ilm pdata.InstrumentationLibraryMetrics) bool {
+			ms := ilm.Metrics()
+			ms.RemoveIf(func(m pdata.Metric) bool {
+				if ctdp.metrics != nil {
+					if _, ok := ctdp.metrics[m.Name()]; !ok {
+						return false
 					}
 				}
-			}
-		}
-	}
+				baseIdentity := tracking.MetricIdentity{
+					Resource:               rm.Resource(),
+					InstrumentationLibrary: ilm.InstrumentationLibrary(),
+					MetricDataType:         m.DataType(),
+					MetricName:             m.Name(),
+					MetricUnit:             m.Unit(),
+				}
+				switch m.DataType() {
+				case pdata.MetricDataTypeSum:
+					ms := m.Sum()
+					if ms.AggregationTemporality() != pdata.AggregationTemporalityCumulative {
+						return false
+					}
+					if ctdp.monotonicOnly && !ms.IsMonotonic() {
+						return false
+					}
+					baseIdentity.MetricIsMonotonic = ms.IsMonotonic()
+					ctdp.convertDataPoints(ms.DataPoints(), baseIdentity)
+					ms.SetAggregationTemporality(pdata.AggregationTemporalityDelta)
+					return ms.DataPoints().Len() == 0
+				default:
+					return false
+				}
+			})
+			return ilm.Metrics().Len() == 0
+		})
+		return rm.InstrumentationLibraryMetrics().Len() == 0
+	})
 	return md, nil
 }
 
 // Shutdown is invoked during service shutdown.
 func (ctdp *cumulativeToDeltaProcessor) Shutdown(context.Context) error {
+	ctdp.cancelFunc()
 	return nil
 }
 
-func newDeltaCalculator() awsmetrics.MetricCalculator {
-	return awsmetrics.NewMetricCalculator(func(prev *awsmetrics.MetricValue, val interface{}, timestamp time.Time) (interface{}, bool) {
-		result := delta{value: val.(float64), prevTimestamp: timestamp}
-
-		if prev != nil {
-			deltaValue := val.(float64) - prev.RawValue.(float64)
-			result.value = deltaValue
-			result.prevTimestamp = prev.Timestamp
-			return result, true
-		}
-		return result, false
-	})
-}
+func (ctdp *cumulativeToDeltaProcessor) convertDataPoints(in interface{}, baseIdentity tracking.MetricIdentity) {
+	switch dps := in.(type) {
+	case pdata.NumberDataPointSlice:
+		dps.RemoveIf(func(dp pdata.NumberDataPoint) bool {
+			id := baseIdentity
+			id.StartTimestamp = dp.StartTimestamp()
+			id.Attributes = dp.Attributes()
+			id.MetricValueType = dp.Type()
+			point := tracking.ValuePoint{
+				ObservedTimestamp: dp.Timestamp(),
+			}
+			if id.IsFloatVal() {
+				// Do not attempt to transform NaN values
+				if math.IsNaN(dp.DoubleVal()) {
+					return false
+				}
+				point.FloatValue = dp.DoubleVal()
+			} else {
+				point.IntValue = dp.IntVal()
+			}
+			trackingPoint := tracking.MetricPoint{
+				Identity: id,
+				Value:    point,
+			}
+			delta, valid := ctdp.deltaCalculator.Convert(trackingPoint)
 
-type delta struct {
-	value         float64
-	prevTimestamp time.Time
+			// When converting non-monotonic cumulative counters,
+			// the first data point is omitted since the initial
+			// reference is not assumed to be zero
+			if !valid {
+				return true
```

Comment:
Sounds good. This behavior is described in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#resets-and-gaps